refactor(trading): pane stores for grid state

mattrussell36 opened the pull request:
# Related issues 🔗

Closes #3396 
Closes #3397 

# Description ℹ️

Adds a way for panes to store state and connects panes that display data grids to store the current filters and sort orders. Going forward we can use these stores to manage other pane based settings.

# Demo 📺

N/A

# Technical 👨‍🔧

- Adds useDataGridEvents hook to connect resize, filter and sort events to a component
- Adds zustand stores for each pane. These are initialized in the app only, not in libs so as not to leak local storage into other apps.
- Remove useBottomPlaceholder hook. For now buttons might overlay the bottom rows, the incoming pane context menu will resolve this.
- Remove useColumnSizes

bglownia commented on apps/trading/components/accounts-container/accounts-container.tsx:
```diff
@@ -75,3 +85,31 @@ export const AccountsContainer = ({
+  update: (gridStore: Store) => void;
+}>()(
+  persist(
+    subscribeWithSelector((set) => ({
```
`subscribeWithSelector` will not give any performance gain if selector returns new array each time.
`Object.is([store.gridStore, store.update], [store.gridStore, store.update])` is always false 

To make it work effectively we can split usage in two hook calls.

```
const gridStore = useAccountStore((store) => store.gridStore);
const update =   useAccountStore((store) => store.update);
```

mattrussell36 replied:
Actually not sure why I'm using subscribeWithSelector. Its not required. Will remove

mattrussell36 replied:
Updated to use different selectors

bglownia commented on apps/trading/components/accounts-container/accounts-container.tsx:
```diff
@@ -75,3 +85,31 @@ export const AccountsContainer = ({
   );
 };
+
+type Store = {
```
Store shape for each table is the same. Do we need to defined it for each table separately?

```
const makeTableStateStore = (name: string) =>
  create<{
    gridStore: Store;
    update: (gridStore: Store) => void;
  }>()(
    persist(
      subscribeWithSelector((set) => ({
        gridStore: {},
        update: (newStore) => {
          set((curr) => ({
            gridStore: {
              ...curr.gridStore,
              ...newStore,
            },
          }));
        },
      })),
      { name }
    )
  );

const useLiquidityStore = makeTableStateStore('vega_ledger_store');
```

bglownia replied:
Store shape is the same as `State` from libs\datagrid\src\lib\use-datagrid-events it will be better if we can import and reuse it

bglownia commented on apps/trading/components/orders-container/orders-container.tsx:
```diff
@@ -0,0 +1,125 @@
+  columnState?: ColumnState[];
+};
+
+const useOrderListStore = create<{
```
We can make it more generic


```
import { immer } from 'zustand/middleware/immer';

const makeMultiinstanceTableStore = (name: string) =>
  create<{
    gridState: Record<string, Store>;
    update: (gridStore: Store, key: string) => void;
  }>()(
    persist(
      subscribeWithSelector(
        immer((set) => ({
          gridState: {},
          update: (newState: Store, key) => {
            set((state) => {
              state.gridState[key] = newState;
            });
          },
        }))
      ),
      {
        name,
      }
    )
  );

const useOrderListStore = makeMultiinstanceTableStore('vega_order_list_store');
```
and then use it in component
```
const StoreKey: Record<Filter, string> = {
  [Filter.Closed]: 'closed',
  [Filter.Open]: 'open',
  [Filter.Rejected]: 'rejected',
};

export const OrdersContainer = ({ marketId, filter }: OrderContainerProps) => {
  [...]
  const storeKey = filter ? StoreKey[filter] : 'all';
  const gridStore = useOrderListStore((store) => store.gridState[storeKey]);
  const update = useOrderListStore((store) => store.update);
  const gridStoreCallbacks = useDataGridEvents(gridStore, (colState) => {
    update(colState, storeKey);
  })
  [...]
}

```

mattrussell36 replied:
I think that each pane should use a single store for all its state, right now that only includes grid state but in future it might also have state for other things (formatting, small balances, chart options, etc etc).

We can use slices (as per zustand documentation) to organise the state and share logic.

bglownia commented:
I'm getting empty rows after applying sort by size in closed orders table and then go to rejected and go back to closed
![image](https://github.com/vegaprotocol/frontend-monorepo/assets/1754247/7af8f8ea-6ba8-4736-8d9d-be22b4199ff1)

mattrussell36 commented:
@bglownia pushed some changes.

- Removed subscribeWithSelector
- Fixed filters. I had actually broken these. The onGridReady function for the order list that set initial filter state was being overridden by the one from useDataGridEvents. I've updated orders-container to use a hook which handles the special cases for the order list
- Added slices to prevent grid store functionality repetition

bglownia commented on apps/trading/components/accounts-container/accounts-container.tsx:
```diff
@@ -75,3 +84,14 @@ export const AccountsContainer = ({
 };
+
+const useAccountStore = create<DataGridSlice>()(
+  persist(
```
it can be just `persist(createDataGridSlice, { name: 'vega_fills_store' })`

mattrussell36 replied:
Done
